iterreg/sparse/solvers.py

```diff
@@ -254,3 +254,42 @@ def fista(X, y, alpha, prox=shrink, pen=ell1, max_iter=1_000, f_store=10,
             if verbose:
                 print(t, E[t // f_store])
     return w, all_w, E
+
+
+@njit
+def adaptive(X, y, alpha, prox=shrink, pen=ell1, max_iter=1_000, n_adapt=2,
+             f_store=10, w_init=None, verbose=False):
+    """Coordinate descent for the Tikhonov problem."""
+    p = X.shape[1]
+    X = np.asfortranarray(X)
+    lc = np.zeros(p)
+    for j in range(p):
+        lc[j] = norm(X[:, j]) ** 2
+
+    E = np.zeros(max_iter * n_adapt // f_store)
+    adapt_coefs = np.ones(p)
+
+    if w_init is None:
+        w = np.zeros(p)
+        R = y.copy().astype(np.float64)
```

Comment:
the astype should not be needed, did you encounter a case where it was ?

Reply:
Can't remember, think not, I remove it